File: solutions/312. Burst Balloons/312-2.py
class Solution:
  def maxCoins(self, nums: list[int]) -> int:
    n = len(nums)
    # dp[i][j] := maxCoins(nums[i..j])
    dp = [[0] * (n + 2) for _ in range(n + 2)]

    nums = [1] + nums + [1]

    for d in range(n):
      for i in range(1, n - d + 1):
        j = i + d
        for k in range(i, j + 1):
          dp[i][j] = max(
              dp[i][j],
              dp[i][k - 1] +
              dp[k + 1][j] +
              nums[i - 1] * nums[k] * nums[j + 1])

    return dp[1][n]

File: solutions/312. Burst Balloons/312-2.java
class Solution {
  public int maxCoins(int[] nums) {
    final int n = nums.length;
    // dp[i][j] := maxCoins(arr[i..j])
    int[][] dp = new int[n + 2][n + 2];
    int[] arr = new int[n + 2];

    System.arraycopy(nums, 0, arr, 1, n);
    arr[0] = 1;
    arr[n + 1] = 1;

    for (int d = 0; d < n; ++d)
      for (int i = 1; i + d <= n; ++i) {
        final int j = i + d;
        for (int k = i; k <= j; ++k)
          dp[i][j] = Math.max(dp[i][j],                     //
                              dp[i][k - 1] + dp[k + 1][j] + //
                                  arr[i - 1] * arr[k] * arr[j + 1]);
      }

    return dp[1][n];
  }
}
